ref: document `exp apply` behavior change (#4349)
iterative/dvc.org: content/docs/command-reference/exp/apply.md

## Synopsis

```usage
usage: dvc exp apply [-h] [-q | -v] experiment

positional arguments:
  experiment     Experiment to be applied
```

## Description

Restores an `experiment` into the workspace. The experiment can be referenced by
name or hash (see `dvc exp run` for details).

Specifically, `dvc exp apply` restores any files or directories which exist in
the experiment, to their exact states from the experiment. This includes files
tracked both with DVC and Git: code, raw data, <abbr>parameters</abbr>,
<abbr>metrics</abbr>, resulting artifacts, etc.

The resulting Git + DVC workspace after `dvc exp apply` will contain the
contents of `experiment`, plus any files from the workspace prior to
`dvc exp apply` that did not exist in `experiment`.

This is typically used after choosing a target `experiment` with `dvc exp show`

> or `dvc exp diff`, and before committing it to Git (making it [persistent].

Reviewer: ```suggestion
or `dvc exp diff`, and before committing it to Git (making it [persistent]).
```

```

## Description

Restores an `experiment` into the workspace. The experiment can be referenced by
name or hash (see `dvc exp run` for details).

Specifically, `dvc exp apply` restores any files or directories which exist in
the experiment, to their exact states from the experiment. This includes files
tracked both with DVC and Git: code, raw data, <abbr>parameters</abbr>,
<abbr>metrics</abbr>, resulting artifacts, etc.

This is typically used after choosing a target `experiment` with `dvc exp show`
or `dvc exp diff`, and before committing it to Git (making it [persistent].

> Note that any [checkpoints] found in the `experiment` will **not** be
> preserved when applying and committing it. Use `dvc exp branch` instead.

<details>

### Expand for `dvc exp apply` implementation details

`dvc exp apply` is approximately equivalent to the following Git + DVC workflow:

```cli
$ git stash --include-untracked
$ git checkout --detach experiment
$ git reset HEAD^1
$ git checkout stash@{0} --ours
$ git stash drop
$ dvc checkout
```

The resulting Git + DVC workspace after `dvc exp apply` will contain the
contents of `experiment`, plus any files from the workspace prior to

> `dvc exp apply` that did not exist in `experiment`.

Reviewer: Thanks for including the clear implementation, but I think it's unnecessary and goes a bit too deep. Can we keep this paragraph in the main text and drop the rest of this details section?

<admon type="info">

Note that `dvc exp apply` does not actually use the standard Git stash (and does
not modify `.git/refs/stash`).

</admon>

</details>

<admon type="warn">

Conflicting changes in the workspace are overwritten, but the result of
`dvc exp apply` can be reverted using Git.

</admon>

<details>

### Expand for details on reverting `dvc exp apply`

`dvc exp apply` can be reverted with the following Git workflow:

```cli

> $ git reset --hard

Reviewer: Why isn't `git stash` sufficient?

Author: Anything equivalent to the hard reset works, whether that's `git stash`, `git checkout --`, etc

Reviewer: Do you mind if we suggest `git stash` then since it's less dangerous?

$ git stash apply refs/exps/apply/stash
```

<admon type="info">

Note that this workflow is only valid as long no Git commands which affect
`HEAD` (such as `git commit` or `git checkout`) have been used since

> `dvc exp apply` was run.

Reviewer: Couldn't you return to that head to recover even if you've done other things?

Author: Yes. You can do

```
$ git reset --hard <original_HEAD>
$ git stash apply refs/exps/apply/stash
```

to recover regardless of what you have done since `dvc exp apply`, but I wasn't sure how much detail we wanted to get into here

Reviewer: Agree about the detail. WDYT about making this slightly less alarming since it sounds like you have no way to recover once you make any changes?

```suggestion
Note that `git stash apply` may fail if you run Git commands which affect
`HEAD` (such as `git commit` or `git checkout`) after
`dvc exp apply`.
```
